src/librustc_mir/borrow_check/mod.rs

```diff
@@ -897,6 +976,55 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                     .enumerate()
                     .for_each(|field| drop_field(self, field));
             }
+
+            // #45696: special-case Box<T> by treating its dtor as
+            // only deep *across owned content*. Namely, we know
+            // dropping a box does not touch data behind any
+            // references it holds (which a Deep Write would touch).
+            ty::TyAdt(def, _) if def.is_box() => {
+                // First, we model the loss of the reference itself
+                // via a shallow write.
+                //
+                // FIXME: Could we just get by with the second step
+                // below? (This step seems like a good thing to do
+                // just as a matter of principle; but none of our
+                // testsuite complained when I tried leaving it out.)
+                self.access_place(
+                    ContextKind::Drop.new(loc),
+                    (drop_place, span),
+                    (ShallowOrDeep::Shallow(None), Write(WriteKind::StorageDeadOrDrop)),
+                    LocalMutationIsAllowed::Yes,
+                    flow_state,
+                );
+
+                // Second, we model the deallocation of the box
+                // contents via a shallow write to a *deref* of the
+                // box, because the shallow write above wouldn't do
+                // such a deref.
+                //
+                // (We use a shallow write because a Deep one would
+                // touch `&mut` references in `T`.)
+                let deref_place = drop_place.clone().deref();
+                self.access_place(
+                    ContextKind::Drop.new(loc),
+                    (&deref_place, span),
+                    (ShallowOrDeep::Shallow(None), Write(WriteKind::StorageDeadOrDrop)),
+                    LocalMutationIsAllowed::Yes,
+                    flow_state,
+                );
+
+                // Third, we model dropping any content owned by the
+                // box by recurring on box contents. This catches
+                // cases like `Box<Box<ScribbleWhenDropped<&mut T>>>`,
+                // while still restricting Write to *owned* content.
```

Comment:
Why is this last? Doesn't dropping the contents happen *before* deallocating the box?

Reply:
If you want the drop of the box contents to happen first, that should be easy to do. For some reason I thought the presentation was more "natural" from the view point of someone reading the code.
 * I didn't think the order in which we did these three steps would actually matter ...
 * (but now you've got me wondering if it could ... I'll do some experiments...)

I do remember playing with moving around the order to see if that got rid of the deltas to `dropck.nll.stderr` and `dropck-object-cycle.nll.stderr`. (But it didn't.)

So, anyway, I'll try putting it first as part of other changes I'm making in response to your feedback.

Reply:
One argument for why the order shouldn't matter here: Given `a_box: Box<T>`, In theory someone should be allowed to move the contents out of `*a_box`, and have those contents get dropped at a time that is disconnected from the (shallow) drop of `a_box`.

Of course that would be totally broken if the drop of `a_box` *did* attempt to dereference and drop the contents at `*a_box`. But that's why the shallow/deep distinction is important here.

Reply:
> (but now you've got me wondering if it could ... I'll do some experiments...)

i played around a little to try to see if I could create a struct with a recursive self-reference inside a box and somehow throw in a destructor that would allow us to in some way observe the access order here. But perhaps unsurprisingly I was unable to make such a thing due to dropck rules...